File: LeetCode/0054-Spiral Matrix/main.cpp
//
// Created by 薛智钧 on 2020/6/5.
//
#include <iostream>
#include <vector>
using namespace std;

/*
 * Given a matrix of m x n elements (m rows, n columns),
 * return all elements of the matrix in spiral order.
 *  Example 2:

Input:
[
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9,10,11,12]
]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 * */

class Solution {
 public:
  static vector<int> spiralOrder(vector<vector<int>>& matrix) {
    int start_row = 0, end_row = matrix.size() - 1;
    int start_col = 0, end_col = matrix[0].size() - 1;
    int dir = 0;
    vector<int> res;

    while (start_row <= end_row && start_col <= end_col) {
      switch (dir) {
        case 0:  // right direction
          for (int c = start_col; c <= end_col; ++c) {
            res.push_back(matrix[start_row][c]);
          }
          ++start_row;
          break;
        case 1:  // down direction
          for (int r = start_row; r <= end_row; ++r) {
            res.push_back(matrix[r][end_col]);
          }
          --end_col;
          break;
        case 2:  // left direction
          for (int c = end_col; c >= start_col; --c) {
            res.push_back(matrix[end_row][c]);
          }
          --end_row;
          break;
        case 3:  // up direction
          for (int r = end_row; r >= start_row; --r) {
            res.push_back(matrix[r][start_col]);
          }
          ++start_col;
          break;
      }
      dir = (dir + 1) % 4;
    }

    return res;
  }
};

int main() {
  vector<vector<int>> matrix{{1, 2, 3}};
  vector<int> res = Solution::spiralOrder(matrix);
  for (int a : res) cout << a << " ";
}

File: LeetCode/0054-Spiral Matrix/main.py
"""  
Given an m x n matrix, return all elements of the matrix in spiral order.

Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
Output: [1,2,3,6,9,8,7,4,5]
"""


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        start_row, end_row = 0, len(matrix) - 1
        start_col, end_col = 0, len(matrix[0]) - 1
        res = []
        dir = 0

        while start_row <= end_row and start_col <= end_col:
            if dir == 0:  # right direction
                for c in range(start_col, end_col + 1):
                    res.append(matrix[start_row][c])
                start_row += 1

            elif dir == 1:  # down direction
                for r in range(start_row, end_row + 1):
                    res.append(matrix[r][end_col])
                end_col -= 1

            elif dir == 2:  # left direction
                for c in range(end_col, start_col - 1, -1):
                    res.append(matrix[end_row][c])
                end_row -= 1

            elif dir == 3:  # up direction
                for r in range(end_row, start_row - 1, -1):
                    res.append(matrix[r][start_col])
                start_col += 1

            dir = (dir + 1) % 4

        return res
